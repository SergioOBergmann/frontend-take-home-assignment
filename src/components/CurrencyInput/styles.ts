import styled from 'styled-components';

export const Icon = styled.div`
  width: 56px;
  height: 56px;

  background: #f4f8fa;
  border: 1px solid #e1e8ed;
  box-sizing: border-box;
  border-radius: 4px 0px 0px 4px;
`;

export const IconSpan = styled.span`
  width: 15px;
  margin: auto;

  font-size: 22px;
  line-height: 26px;

  letter-spacing: -0.183333px;

  color: #657786;
`;

export const Input = styled.input`
  flex: 1;
  width: 100%;
  height: 56px;

  padding: 16px;

  background: #ffffff;
  border: 1px solid #e1e8ed;
  border-left: 0;
  box-sizing: border-box;
  border-radius: 0px 4px 4px 0px;

  font-weight: 600;
  font-size: 20px;

  letter-spacing: -0.166667px;
`;
