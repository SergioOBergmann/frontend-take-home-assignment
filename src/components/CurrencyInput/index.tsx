import * as React from 'react';

import { Icon, IconSpan, Input } from './styles';

import { InputDiv, InputLabel, LOCALE } from '../../styles/global';

type CurrencyInputProps = {
  label: string;
  currencySymbol?: string;
  defaultValue: string;
  onChange: (value: string) => void;
};

const CurrencyInput = ({
  label,
  currencySymbol = '$',
  defaultValue,
  onChange
}: CurrencyInputProps): JSX.Element => {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    let totalAmount = +event.target.value.replace(/[, ]+/g, '');
    if (
      isNaN(totalAmount) ||
      event.target.value.length >= 22 ||
      (!isNaN(totalAmount) && totalAmount < 0)
    ) {
      event.target.value = event.target.defaultValue;
    } else if (event.target.value.indexOf('.') !== -1) {
      const totalAmountText = event.target.value.replace(/[, ]+/g, '');
      const numberParts = totalAmountText.split('.');
      const integer = numberParts[0];
      const fraction = numberParts[1];
      if (fraction.length > 2) {
        totalAmount = +totalAmountText.replace('.', '');
        totalAmount = totalAmount / 100;
      } else if (fraction.length === 1) {
        totalAmount = +totalAmountText.replace('.', '');
        totalAmount = totalAmount / 100;
      }
      event.target.value = totalAmount.toLocaleString(LOCALE, {
        minimumFractionDigits: 2
      });

      if (
        event.target.value.split('.')[1].length == 2 &&
        event.target.value.split('.')[1] === '00'
      ) {
        event.target.setSelectionRange(
          event.target.value.length - 2,
          event.target.value.length
        );
      }
      onChange(event.target.value);
    } else {
      event.target.value = totalAmount.toLocaleString(LOCALE);
      onChange(event.target.value);
    }
  };

  return (
    <>
      <InputDiv>
        <InputLabel>{label}</InputLabel>
        <Icon>
          <IconSpan>{currencySymbol}</IconSpan>
        </Icon>
        <Input
          data-testid="currency-input"
          defaultValue={defaultValue}
          onChange={handleChange}
        />
      </InputDiv>
    </>
  );
};

export default CurrencyInput;
