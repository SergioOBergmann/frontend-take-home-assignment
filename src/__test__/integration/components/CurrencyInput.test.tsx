import * as React from 'react';

import {
  render,
  RenderResult,
  fireEvent,
  screen
} from '@testing-library/react';

import CurrencyInput from '../../../components/CurrencyInput';

let documentBody: RenderResult;
let input: HTMLElement;

describe('<CurrencyInput />', () => {
  let onChange = jest.fn();
  beforeEach(() => {
    onChange = jest.fn();
    documentBody = render(
      <CurrencyInput
        label="Total amount"
        defaultValue="25,000"
        onChange={onChange}
      />
    );
    input = screen.getByTestId('currency-input');
  });

  it('matches snapshot', () => {
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  describe('handle change for valid values', () => {
    it('when an integer value greater than a thousand has been inputted', () => {
      fireEvent.change(input, { target: { value: '1001' } });

      expect(input).toHaveValue('1,001');
      expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('when a decimal value with 2 fraction digits has been inputted', () => {
      fireEvent.change(input, { target: { value: '1001.52' } });

      expect(input).toHaveValue('1,001.52');
      expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('when a decimal value with 1 fraction digit has been inputted', () => {
      fireEvent.change(input, { target: { value: '1001.5' } });

      expect(input).toHaveValue('100.15');
      expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('when a decimal value with 1 fraction non-significant digit has been inputted', () => {
      fireEvent.change(input, { target: { value: '1001.0' } });

      expect(input).toHaveValue('10,010.00');
      expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('when a decimal value with 3 fractions digits has been inputted', () => {
      fireEvent.change(input, { target: { value: '1001.523' } });

      expect(input).toHaveValue('10,015.23');
      expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('when a dot has been typed at the last position of the value', () => {
      fireEvent.change(input, { target: { value: '1001.' } });

      expect(input).toHaveValue('1,001.00');
      expect(onChange).toHaveBeenCalledTimes(1);
    });
  });

  describe('handle change for invalid integer values', () => {
    it('when a non-numeric char has been typed', () => {
      fireEvent.change(input, { target: { value: '25,000a' } });

      expect(input).toHaveValue('25,000');
      expect(onChange).not.toHaveBeenCalled();
    });
  });
});

describe('<CurrencyInput />', () => {
  let onChange = jest.fn();
  beforeEach(() => {
    onChange = jest.fn();
    documentBody = render(
      <CurrencyInput
        label="Total amount"
        defaultValue="25,000.59"
        onChange={onChange}
      />
    );
    input = screen.getByTestId('currency-input');
  });

  it('handle change for invalid decimal values > when a non-numeric char has been typed', () => {
    fireEvent.change(input, { target: { value: '25,000.59a' } });

    expect(input).toHaveValue('25,000.59');
    expect(onChange).not.toHaveBeenCalled();
  });

  it('handle change for invalid values > when more than one decimal separator has been typed', () => {
    fireEvent.change(input, { target: { value: '25000.59.' } });

    expect(input).toHaveValue('25,000.59');
    expect(onChange).not.toHaveBeenCalled();
  });
});
