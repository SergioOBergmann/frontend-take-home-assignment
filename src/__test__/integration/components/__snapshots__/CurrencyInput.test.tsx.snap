// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CurrencyInput /> matches snapshot 1`] = `
<body>
  <div>
    <div
      class="sc-gzVnrw eDRxMu"
    >
      <label
        class="sc-htoDjs fTBVLS"
      >
        Total amount
      </label>
      <div
        class="sc-bdVaJa gbgDHS"
      >
        <span
          class="sc-bwzfXH gvUvYq"
        >
          $
        </span>
      </div>
      <input
        class="sc-htpNat ipwnkg"
        data-testid="currency-input"
        value="25,000"
      />
    </div>
  </div>
</body>
`;
